src/basic/virt.c

```diff
@@ -294,6 +294,11 @@ static Virtualization detect_vm_dmi(void) {
               }
         }
 
+        /* The DMI vendor tables in /sys/class/dmi/id don't help us distinguish between GCE
+         * virtual machines and bare-metal instances, so we need to look at supervisor. */
+        if (r == VIRTUALIZATION_GOOGLE && detect_vm_cpuid() != VIRTUALIZATION_KVM)
```

Comment:
The check should be inverted. Instead of assuming that there is no virtualization in case the virt stack is not exactly as expected, we should instead add a check that the virt stack is as expected in the virtualized case.
In particular, we should fall back to the SMBIOS_VM_BIT_SET check below, so that we return something useful if later on the stack is fixed to set the bit.

Reply:
Indeed you are right, it makes more sense to reverse the condition, I will make the change. Thanks for the feedback!

Reply:
Maybe it would be better to not change the condition but instead of returning VIRTUALIZATION_NONE, I will change the value of `r=VIRTUALIZATION_NONE.` To be able to fall back in the `SMBIOS_VM_BIT_SET` check below if the condition is true meaning the virtualization wasn't as expected.

In that way, if the condition fails, it would mean that the virt stack is as expected, because `r=VIRTUALIZATION_GOOGLE`  and KVM has been detected. Therefore function's return will keep the value of r.

Does this sound better to you?

Reply:
This still has the original problem that it checks for `detect_vm_cpuid() != VIRTUALIZATION_KVM`. We want to treat the case of `r == VIRTUALIZATION_GOOGLE && detect_vm_cpuid() == VIRTUALIZATION_KVM` as special.

Reply:
Understood, I made the change accordingly :)

Reply:
Or in other words. Let's say that we get `r == VIRTUALIZATION_GOOGLE`, but `detect_vm_cpuid()` returns `VIRTUALIZATION_GNX`. I don't think we should downgrade that to `VIRTUALIZATION_NONE`.

Reply:
Yes, I wanted to have a similar output as previously. Keeping _google_ as output of `systemd-detect-virt` if it runs in a GCE VM, and print _none_ if we are in a bare-metal machine.